rust_icu_unorm2/src/lib.rs

```diff
@@ -0,0 +1,150 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Contains implementations of functions from ICU's `unorm2.h`.
+
+use {
+    rust_icu_common as common, rust_icu_sys as sys, rust_icu_sys::versioned_function,
+    rust_icu_sys::*, rust_icu_ustring as ustring,
+    rust_icu_ustring::buffered_uchar_method_with_retry,
+};
+use std::convert::{TryFrom, TryInto};
+
+#[derive(Debug)]
+pub struct UNormalizer {
+    rep: std::ptr::NonNull<sys::UNormalizer2>,
+    owned: bool,
+}
+
+impl Drop for UNormalizer {
+    /// Implements `unorm2_close`
+    fn drop(&mut self) {
+        // Close the normalizer only if it is unowned.
+        if self.owned {
+            unsafe {
+                versioned_function!(unorm2_close)(self.rep.as_ptr())
+            }
+        }
+    }
+}
+
+impl UNormalizer {
+    /// Implements `unorm2_getNFCInstance`.
+    pub fn new_nfc() -> Result<Self, common::Error> {
+        unsafe { UNormalizer::new_normalizer_unowned(versioned_function!(unorm2_getNFCInstance)) }
+    }
+
+    /// Implements `unorm2_getNFDInstance`.
+    pub fn new_nfd() -> Result<Self, common::Error> {
+        unsafe { UNormalizer::new_normalizer_unowned(versioned_function!(unorm2_getNFDInstance)) }
+    }
+
+    /// Implements `unorm2_getNFKCInstance`.
+    pub fn new_nfkc() -> Result<Self, common::Error> {
+        unsafe { UNormalizer::new_normalizer_unowned(versioned_function!(unorm2_getNFKCInstance)) }
+    }
+
+    /// Implements `unorm2_getNFKDInstance`.
+    pub fn new_nfkd() -> Result<Self, common::Error> {
+        unsafe { UNormalizer::new_normalizer_unowned(versioned_function!(unorm2_getNFKDInstance)) }
+    }
+
+    /// Implements `unorm2_getNFKCCasefoldInstance`.
+    pub fn new_nfkc_casefold() -> Result<Self, common::Error> {
+        unsafe { UNormalizer::new_normalizer_unowned(versioned_function!(unorm2_getNFKCCasefoldInstance)) }
+    }
+
+    unsafe fn new_normalizer_unowned(
+        constrfn: unsafe extern "C" fn(*mut sys::UErrorCode) -> *const sys::UNormalizer2) -> Result<Self, common::Error> {
+        let mut status = common::Error::OK_CODE;
+        let rep = {
+            assert!(common::Error::is_ok(status));
+            let ptr = constrfn(&mut status) as *mut sys::UNormalizer2;
+            std::ptr::NonNull::new_unchecked(ptr)
+        };
+        common::Error::ok_or_warning(status)?;
+        Ok(UNormalizer{ rep, owned: false })
+    }
+
+    /// Implements `unorm2_normalize`.
+    pub fn normalize(&self, norm: &str) -> Result<String, common::Error> {
+        let norm = ustring::UChar::try_from(norm)?;
+        let result = self.normalize_ustring(&norm)?;
+        String::try_from(&result)
+    }
+
+    /// Implements `unorm2_normalize`.
+    pub fn normalize_ustring(
+        &self,
+        norm: &ustring::UChar
+        ) -> Result<ustring::UChar, common::Error> {
+        const CAPACITY: usize = 200;
```

Comment:
Could we take a better guess here? We have the string to be normalized, so maybe some multiple of its length?

Reply:
We possibly could. However, allowing non-`const` here would require reworking the macro, which I am reluctant to do for unclear gain.